src/Microsoft.Health.Fhir.Shared.Core/Features/Search/SearchIndexer.cs

```diff
@@ -234,12 +234,32 @@ private IEnumerable<SearchIndexEntry> ProcessNonCompositeSearchParameter(SearchP
                     continue;
                 }
 
-                _logger.LogDebug(
-                    "The FHIR element '{ElementType}' will be converted using '{ElementTypeConverter}'.",
-                    extractedValue.TypeName,
-                    converter.GetType().FullName);
+                IEnumerable<ISearchValue> searchValues = converter.ConvertTo(extractedValue);
 
-                results.AddRange(converter.ConvertTo(extractedValue) ?? Enumerable.Empty<ISearchValue>());
+                if (searchValues != null)
+                {
+                    if (searchParameterType == SearchParamType.Reference && allowedReferenceResourceTypes?.Count == 1)
+                    {
+                        // For references, if the type is not specified in the reference string, we can set the type on the search value because
+                        // in this case it can only be of one type.
+                        string singleAllowedResourceType = null;
```

Comment:
>null [](start = 59, length = 4)

Are we doing late binding via `??==` because of optimization of the case when we don't have ReferenceSearchValue in searchValues and allowedReferenceResourceTypes.Single() is costly operation?

Reply:
Actually, I put the Single() call in before switching to IReadOnlyCollection, so this is not necessary anymore. Thank you for catching that.